src/infra/dex/balancer/dto.rs

```diff
@@ -1,56 +1,243 @@
+        tokenOut: $tokenOut,
+        useProtocolVersion: $useProtocolVersion
+    ) {
+        tokenAddresses
+        swaps {
+            poolId
+            assetInIndex
+            assetOutIndex
+            amount
+            userData
+        }
+        swapAmountRaw
+        returnAmountRaw
+        tokenIn
+        tokenOut
+    }
+}
+"#;
+
+#[derive(Serialize)]
 #[serde(rename_all = "camelCase")]
-pub enum OrderKind {
-    Sell,
-    Buy,
+pub struct Query<'a> {
+    query: &'a str,
+    variables: Variables,
 }
 
-/// An SOR query.
+impl Query<'_> {
+    pub fn from_domain(
+        order: &dex::Order,
+        tokens: &auction::Tokens,
+        slippage: &dex::Slippage,
+        chain_id: eth::ChainId,
+        contract_address: eth::ContractAddress,
+        query_batch_swap: bool,
+        swap_deadline: Option<u64>,
+    ) -> Result<Self, Error> {
+        let token_decimals = match order.side {
+            order::Side::Buy => tokens
+                .decimals(&order.buy)
+                .ok_or(Error::MissingDecimals(order.buy)),
+            order::Side::Sell => tokens
+                .decimals(&order.sell)
+                .ok_or(Error::MissingDecimals(order.sell)),
+        }?;
+        let variables = Variables {
+            call_data_input: CallDataInput {
+                deadline: swap_deadline,
+                receiver: contract_address.0,
+                sender: contract_address.0,
+                slippage_percentage: slippage.as_factor().clone(),
+            },
+            chain: Chain::from_domain(chain_id)?,
+            query_batch_swap,
+            swap_amount: HumanReadableAmount::from_decimal_units(
+                &order.amount.get(),
+                token_decimals,
+            ),
+            swap_type: SwapType::from_domain(order.side),
+            token_in: order.sell.0,
+            token_out: order.buy.0,
+            use_protocol_version: Some(VaultVersion::V2.into()),
+        };
+        Ok(Self {
+            query: QUERY,
+            variables,
+        })
+    }
+}
+
+/// Refers to the SOR API V3's `AmountHumanReadable` type and represents a token
+/// amount without decimals.
 #[serde_as]
-#[derive(Debug, Serialize)]
-#[serde(rename_all = "camelCase")]
-pub struct Query {
-    /// The sell token to quote.
-    pub sell_token: H160,
-    /// The buy token to quote.
-    pub buy_token: H160,
-    /// The order kind to use.
-    pub order_kind: OrderKind,
-    /// The amount to quote
-    ///
-    /// For sell orders this is the exact amount of sell token to trade, for buy
-    /// orders, this is the amount of buy tokens to buy.
-    #[serde_as(as = "serialize::U256")]
-    pub amount: U256,
-    /// The current gas price estimate used for determining how the trading
-    /// route should be split.
-    #[serde_as(as = "serialize::U256")]
-    pub gas_price: U256,
+#[derive(Debug, Default, PartialEq)]
+pub struct HumanReadableAmount {
+    value: BigDecimal,
+    decimals: BigDecimal,
```

Comment:
I think the proper types would be `value: U256` and `decimals: u8`.
Having reasonable types here is probably more important than avoiding conversions in functions that get called rarely anyway.

Reply:
That would require converting the values to BigDecimal each time. That is why I decided to keep BigDecimal.